### ISSUE TITLE
body transformation middleware not working

### ISSUE THREAD

**Issue by dotsource** [microsoft/reverse-proxy#899]
I have a middleware that replaces the body like below for example but the ContentLength of the request is not getting updated and proxy is not taking in the new body. Any idea what I should be doing in the body transform middleware to let proxy know that body is changed?

```c#
var ms = new MemoryStream(Encoding.UTF8.GetBytes("test"));
context.Request.Body = ms;
```
### To Reproduce

Create a custom middleware and change the body and its is not being picked up by proxy.

**Comment by Tratcher** [comment]
Can you give a bit more context for your middleware?

ContentLength is something you need to modify yourself, that's not automatically generated.

Are you using .NET 5 or 3.1?

Due to some quirks in HttpClient, HttpProxy is very careful to check if the request declares a body or not. Check out:
https://github.com/microsoft/reverse-proxy/blob/6918b54ea8e25ac0e1d0d81eee871a332387616e/src/ReverseProxy/Service/Proxy/HttpProxy.cs#L317-L361

If you're adding a body to a request that did not previously have one you may need to update some of these related fields so it will be detected.

**Comment by Tratcher** [comment]
Hmm, it would help if you shared the actual middleware code.

**Comment by dotsource** [comment]
> Hmm, it would help if you shared the actual middleware code.

Thanks for your time and help.

               ```context.Request.EnableBuffering();
                var buffer = new byte[Convert.ToInt32(context.Request.ContentLength)];
                await context.Request.Body.ReadAsync(buffer, 0, buffer.Length);
                var bodyAsText = Encoding.UTF8.GetString(buffer);
                context.Request.Body.Position = 0;               

                var newParamValue = $@"test";                
                var qs = HttpUtility.ParseQueryString(bodyAsText);                
                qs["newParam"] = newParamValue ;                

                var ms = new MemoryStream(Encoding.UTF8.GetBytes(qs.ToString()));
                context.Request.Body = ms;```

**Comment by Tratcher** [comment]
Some fixups:
- You don't need EnableBuffering, you have your own buffer.
- ReadAsync isn't adequate, it may only return part of the body, you need to read all the way to the end.
- Reset the ContentLength.
```
    var reader = new StreamReader(context.Request.Body);
    var bodyAsText = await reader.ReadToEndAsync();

    var newParamValue = $@"test";
    var qs = HttpUtility.ParseQueryString(bodyAsText);
    qs["newParam"] = newParamValue;

    var form = qs.ToString();
    var ms = new MemoryStream(Encoding.UTF8.GetBytes(form));
    context.Request.ContentLength = ms.Length;
    context.Request.Body = ms;
    await next();
```

**Comment by dotsource** [comment]
> Some fixups:
> 
> * You don't need EnableBuffering, you have your own buffer.
> * ReadAsync isn't adequate, it may only return part of the body, you need to read all the way to the end.
> * Reset the ContentLength.
> 
> ```
>     var reader = new StreamReader(context.Request.Body);
>     var bodyAsText = await reader.ReadToEndAsync();
> 
>     var newParamValue = $@"test";
>     var qs = HttpUtility.ParseQueryString(bodyAsText);
>     qs["newParam"] = newParamValue;
> 
>     var form = qs.ToString();
>     var ms = new MemoryStream(Encoding.UTF8.GetBytes(form));
>     context.Request.ContentLength = ms.Length;
>     context.Request.Body = ms;
>     await next();
> ```

Thanks for the fixups. Unfortunately the changes did not make any difference and the behavior is the same. ContentLength assignment above did not change the old value!

**Comment by Tratcher** [comment]
It's working fine in the app I've tested. At this point it seems like a contextual problem, e.g. the code is OK but it's in the wrong place. Here's the sample I copied that code into, and where I put it:
https://github.com/microsoft/reverse-proxy/blob/e9255a8e57518821f4eac2d5fab6c5be60fa7c96/testassets/ReverseProxy.Config/Startup.cs#L48-L67

If yours logic is going somewhere else please share your whole Startup.cs and any other relevant classes.

**Comment by dotsource** [comment]
> It's working fine in the app I've tested. At this point it seems like a contextual problem, e.g. the code is OK but it's in the wrong place. Here's the sample I copied that code into, and where I put it:
> https://github.com/microsoft/reverse-proxy/blob/e9255a8e57518821f4eac2d5fab6c5be60fa7c96/testassets/ReverseProxy.Config/Startup.cs#L48-L67
> 
> If yours logic is going somewhere else please share your whole Startup.cs and any other relevant classes.

Thanks for helping to troubleshoot.

I don't see any middleware code in your sample. I am using LoadFromMemory(GetRoutes(), GetClusters()) sample and in Configure, I am just using (nothing inside the 
 app.UseEndpoints(endpoints =>
            {
                endpoints.MapReverseProxy();
            });

The middleware I have is just called above this block use the app.usemiddleware<>(). I was thinking that the middleware should not have anything to do with the proxy and gets executed before the proxy picks up the request and it seems it does. Here is the startup code. It will really help if you show where you put the middleware code in the Configure method.

```
public class Startup
    {
       

        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            // Add the reverse proxy to capability to the server
            var proxyBuilder = services.AddReverseProxy();
            // Initialize the reverse proxy from the "ReverseProxy" section of configuration
            //proxyBuilder.LoadFromConfig(Configuration.GetSection("ReverseProxy"));
            proxyBuilder
                .LoadFromMemory(GetRoutes(), GetClusters())                
                .AddTransforms(transformBuilderContext =>
                {
                    /*// For each route+cluster pair decide if we want to add transforms, and if so, which?
                    // This logic is re-run each time a route is rebuilt.

                    transformBuilderContext.AddPathPrefix("/prefix");

                    // Only do this for routes that require auth.
                    if (string.Equals("token", transformBuilderContext.Route.AuthorizationPolicy))
                    {
                        transformBuilderContext.AddRequestTransform(async transformContext =>
                        {
                            // AuthN and AuthZ will have already been completed after request routing.
                            var ticket = await transformContext.HttpContext.AuthenticateAsync("token");
                            var tokenService = transformContext.HttpContext.RequestServices.GetRequiredService<TokenService>();
                            var token = await tokenService.GetAuthTokenAsync(ticket.Principal);
                            transformContext.ProxyRequest.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
                        });
                    }*/
                }); 
          
            services.AddAuthentication(options =>
            {
                options.DefaultAuthenticateScheme = OktaDefaults.ApiAuthenticationScheme;
                options.DefaultChallengeScheme = OktaDefaults.ApiAuthenticationScheme;
                options.DefaultSignInScheme = OktaDefaults.ApiAuthenticationScheme;
            })
            .AddOktaWebApi(new OktaWebApiOptions()
            {
                OktaDomain = Configuration["Okta:OktaDomain"],
            });          

            services.AddAuthorization();

            services.AddControllers();
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            app.UseHttpsRedirection();

            app.UseRouting();

            //app.UseCors("AllowAll");

            app.UseAuthentication();

            app.UseAuthorization();
			
			
			//this is the custom middleware that changes the request body
            app.UseMiddleware<RequestsMiddleware>();            

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllers();
                endpoints.MapReverseProxy(/*proxyPipeline =>
                {
                    
                }*/);
            });
        }

        //**
        private IReadOnlyList<ProxyRoute> GetRoutes()
        {
            var proxyRouteStatic = new ProxyRoute()
            {
                RouteId = "route1",
                ClusterId = "cluster1",
                Match = new ProxyMatch
                {
                    Path = "/somepath/{**catch-all}"
                }
            };
            var proxyRouteMapServer = new ProxyRoute()
            {
                RouteId = "route2",
                ClusterId = "cluster1",
                Match = new ProxyMatch
                {
                    Path = "/someotherpath/{**catch-all}",
                },
            }.WithTransformPathRouteValues("/someotherpath/{**catch-all}");


            return new List<ProxyRoute>
            {
                proxyRouteStatic,
                proxyRouteMapServer
            };
        }

        private IReadOnlyList<Cluster> GetClusters()
        {
            return new List<Cluster>
            {
                new Cluster
                {
                    Id = "cluster1",
                    Destinations = new Dictionary<string, Destination>
                    {
                        { "destination1", new Destination { Address = "https://destination.com" } }
                    }
                }
            };
        }

        //**
    }

```

**Comment by Tratcher** [comment]
I had been testing with Kestrel, but switching to IIS Express I can also reproduce this. Setting the Content-Length header on the HttpRequest and then enumerating the request headers causes it to return two Content-Length headers, the new one and the old one.

Workaround:
```diff
- context.Request.ContentLength = ms.Length;
+ context.Request.Headers[HeaderNames.ContentLength] = ms.Length.ToString(CultureInfo.InvariantCulture);
```

I'll get this filed in dotnet/AspNetCore.

**Comment by karelz** [comment]
Triage: Root cause was fixed in 6.0 - https://github.com/dotnet/aspnetcore/issues/31735
Workaround above should work for now. Closing.

**Comment by dotsource** [comment]
Thanks for fixing this quick!
